File: contract.go
package memory

import "time"

// Store is the DRAFT API contract — the reviewable shape of the library.
// Phase 1 (docs/MIGRATION.md) replaces this interface with a concrete
// *Store over two SQLite files (memory.db + a disposable index.db); the
// interface exists now so the surface can be criticized before it is load-
// bearing. Method-set changes before Phase 1 are free; after, they cost.
type Store interface {
	// --- the spine ---

	// CreateNode writes an owner-authored node (born active) with provenance.
	CreateNode(n Node, p Provenance) (Node, error)
	// GetNode fetches by id regardless of status (hosts gate display).
	GetNode(id string) (Node, error)
	// UpdateNode edits an ACTIVE owner-authored node in place.
	UpdateNode(id string, title, body *string, props map[string]any) (Node, error)
	// Link writes an edge (idempotent on source+target+type).
	Link(source, target, edgeType, context string) (Edge, error)
	// Neighborhood returns the 1-hop active set around a node.
	Neighborhood(id string) ([]Node, error)

	// --- the consent boundary ---

	// Propose runs the write-time adjudication gate and, unless the gate
	// no-ops, parks the item in the consent queue.
	Propose(p Proposal) (Node, Outcome, error)
	// ProposeEdit parks a change to an active node without applying it.
	ProposeEdit(id string, fields map[string]string, archive bool, p Provenance) error
	// PendingQueue lists everything awaiting the owner, with conflict hints.
	PendingQueue() ([]Pending, error)
	// Decide applies the owner's verdict; compound decisions (supersede)
	// commit their whole sequence and audit every step.
	Decide(id string, d Decision) (Node, error)

	// --- recall ---

	// Recall is ranked retrieval over active, surfaceable nodes of one type:
	// FTS5 lexical relevance blended with recency, importance, and
	// reinforcement; fused with embedder cosine when an Embedder is set.
	Recall(terms []string, nodeType string, limit int) ([]Node, error)
	// Search is cross-type recall over all active, surfaceable knowledge.
	Search(terms []string, limit int) ([]Node, error)
	// Touch records that recalled knowledge was actually used.
	Touch(id string) error

	// --- lifecycle ---

	// Transition moves a node through the status FSM (owner action).
	Transition(id string, to Status) (Node, error)
	// SetSurfacing sets the always/ask/never axis on a node.
	SetSurfacing(id string, s Surfacing) error
	// Quarantine suppresses a node everywhere with an optional review date.
	Quarantine(id string, reviewAt *time.Time) error
	// Forget runs the honest cascade: tombstone the node, drop its edges,
	// scrub indexes, flag derived artifacts stale via lineage, and write a
	// content-free audit entry. It returns what it could NOT reach (e.g.
	// prose mentions needing owner review) rather than overclaiming.
	Forget(id string) (ForgetReport, error)

	// --- lineage & measurement ---

	// RecordDerivation registers a derived artifact's sources (lineage).
	RecordDerivation(artifactID string, sourceIDs []string) error
	// StaleDerivations lists derived artifacts whose sources changed.
	StaleDerivations() ([]string, error)
	// Doctor computes the metadata-only health report. It reports; it
	// never acts.
	Doctor(now time.Time) (DoctorReport, error)
}

// ForgetReport is the honest account of a Forget cascade: what was
// tombstoned, what was scrubbed, and what needs the owner (prose mentions,
// exports already written).
type ForgetReport struct {
	Tombstoned    []string // node ids content-destroyed
	EdgesDropped  int
	IndexScrubbed bool
	FlaggedStale  []string // derived artifacts marked stale via lineage
	NeedsOwner    []string // refs the cascade cannot honestly resolve alone
}

// DoctorReport is the metadata-only health snapshot; every field is
// computable without a model call and phrased as candidates, never actions.
type DoctorReport struct {
	ActiveCount, PendingCount int
	AcceptRate30d             float64  // owner decisions: approved / decided
	DeadWeightCandidates      []string // active, never recalled, aged — review, don't auto-archive
	StaleCandidates           []string // possibly superseded by later activity
	DuplicateCandidates       [][2]string
	QueueOldestDays           int
}

File: src/contract.ts
/**
 * The DRAFT Store contract — the reviewable shape of the library. Phase 1
 * (docs/MIGRATION.md) replaces this interface with the concrete Store class
 * over memory.db + index.db; it exists now so the surface can be criticized
 * before it is load-bearing. Changes are free until Phase 1 ships.
 *
 * Everything is SYNCHRONOUS (DESIGN.md): bun:sqlite is sync, personal scale
 * is sub-millisecond, and the one truly async concern — embedding text —
 * lives in hosts. Vectors in, never models.
 */

import type { Conflict, Decision, Outcome, Pending, Proposal } from "./consent.ts";
import type { ForgetReport } from "./lifecycle.ts";
import type { Edge, Node, NodeId, NodeTypeSpec, Props, Status, Surfacing } from "./types.ts";

/** Tunables for the recall ranking blend; conformance pins the defaults. */
export interface RankingConfig {
  readonly lambda: number; // recency decay per day (importance-dampened)
  readonly reinforcement: number; // use_count weight (default 0.2)
  readonly rrfK: number; // reciprocal-rank fusion constant (default 60)
}

export interface RecallOptions {
  readonly type?: string; // restrict to one node type (e.g. "memory")
  readonly limit?: number; // default 8
  /** Host-embedded query vector: enables cosine fusion over putVector data. */
  readonly queryVector?: Float32Array;
  /** Vector-space identity; required with queryVector. */
  readonly model?: string;
}

/** Metadata-only health snapshot — candidates, never actions. */
export interface DoctorReport {
  readonly activeCount: number;
  readonly pendingCount: number;
  readonly acceptRate30d: number | null; // null: no decisions in window
  readonly deadWeightCandidates: readonly NodeId[]; // dormant ≠ dead — review only
  readonly staleCandidates: readonly NodeId[];
  readonly duplicateCandidates: ReadonlyArray<readonly [NodeId, NodeId]>;
  readonly queueOldestDays: number | null;
}

/** The draft contract. Phase 1 ships `class Store implements StoreContract`. */
export interface StoreContract {
  // --- the spine ---

  /** Register or update a node type (I1: bornStatus is the consent split). */
  registerType(spec: NodeTypeSpec): void;
  /** Owner-authored write — born active, provenance mandatory (I10). */
  createNode(input: {
    type: string;
    title: string;
    body?: string;
    props?: Props;
    importance?: number;
    surfacing?: Surfacing;
    origin: string;
    author?: string;
  }): Node;
  /** Fetch by id regardless of status — hosts gate display. */
  getNode(id: NodeId): Node;
  /** Edit an ACTIVE owner-authored node in place. */
  updateNode(id: NodeId, patch: { title?: string; body?: string; props?: Props }): Node;
  /** Idempotent on (source, target, type). */
  link(source: NodeId, target: NodeId, type?: string, context?: string): Edge;
  /** 1-hop active set (I3). */
  neighborhood(id: NodeId): Node[];

  // --- the consent boundary ---

  /** The write-time gate (I4): created | merged_pending | exists_active. */
  propose(p: Proposal): Outcome;
  /** Park a change to an active node without applying it. */
  proposeEdit(
    id: NodeId,
    change: { fields?: Record<string, string>; archive?: boolean; origin: string; author?: string },
  ): void;
  /** Everything awaiting the owner, oldest first, with conflict hints. */
  pendingQueue(): Pending[];
  /** Apply the owner's verdict; compound verdicts run ordered + audited (I5). */
  decide(id: NodeId, decision: Decision): Node;
  /** Recompute hints for one pending item (also embedded in pendingQueue). */
  conflictsFor(id: NodeId): Conflict[];

  // --- recall ---

  /** Ranked retrieval over active, surfaceable nodes (I2): FTS × recency ×
   * importance × reinforcement; RRF-fused with cosine when a queryVector is
   * supplied. Deterministic without one — and that is not a degraded mode. */
  recall(terms: readonly string[], opts?: RecallOptions): Node[];
  /** Cross-type recall over all active, surfaceable knowledge. */
  search(terms: readonly string[], limit?: number): Node[];
  /** Record that recalled knowledge was actually used (feeds ranking + doctor). */
  touch(id: NodeId): void;

  // --- lifecycle ---

  /** Move through the status FSM (owner action; validates I8 terminality). */
  transition(id: NodeId, to: Status): Node;
  setSurfacing(id: NodeId, s: Surfacing): void;
  /** Suppress everywhere, ask-twice to view, optional re-review date. */
  quarantine(id: NodeId, reviewAt?: string): void;
  /** The honest erasure cascade (I6/I7). */
  forget(id: NodeId): ForgetReport;

  // --- identity, phase A: names (docs/ENTITIES.md) ---

  /** Record a name the node also answers to (owner verb; active nodes;
   * idempotent; audited content-free — the alias text never enters the log). */
  addAlias(id: NodeId, alias: string): void;
  removeAlias(id: NodeId, alias: string): void;
  /** All names the node answers to (normalized), alphabetical. */
  aliasesOf(id: NodeId): string[];
  /** Who is "Ana"? Exact-normalized candidates within one type — the owner
   * picks, the library never does. I2: never invisible, ask resolves. */
  resolveRef(type: string, text: string): Node[];
  /** Walk merged_into chains to the living end; non-merged returns itself. */
  survivorOf(id: NodeId): Node;
  /** Deterministic candidate generation (R1 title, R2 token-subset, R3
   * alias) writing identity questions to the queue — owner/host-scheduled,
   * never ambient. Returns questions added (≤ cap). */
  suggestIdentities(type: string, cap?: number): number;

  // --- lineage & vectors & measurement ---

  /** Register a derived artifact's sources (artifact/source: node id or host ref). */
  recordDerivation(artifact: string, sources: readonly string[]): void;
  /** Derived artifacts whose sources changed or were forgotten. */
  staleDerivations(): string[];
  /** Maintain the vector sidecar — host-computed vectors only. */
  putVector(id: NodeId, model: string, vec: Float32Array): void;
  deleteVectors(model?: string): void;
  /** Rebuild index.db from memory.db (I13 — always safe, always exact). */
  rebuildIndex(): void;
  /** Metadata-only health report — reports, never acts. */
  doctor(now?: Date): DoctorReport;
}
